weed/wdclient/masterclient.go

```diff
@@ -220,39 +235,45 @@ func (mc *MasterClient) tryConnectToMaster(master pb.ServerAddress) (nextHintedL
 		mc.setCurrentMaster(master)
 
 		for {
-			resp, err := stream.Recv()
-			if err != nil {
-				glog.V(0).Infof("%s.%s masterClient failed to receive from %s: %v", mc.FilerGroup, mc.clientType, master, err)
-				stats.MasterClientConnectCounter.WithLabelValues(stats.FailedToReceive).Inc()
-				return err
-			}
+			select {
+			case <-ctx.Done():
+				glog.V(0).Infof("Connection attempt to master stopped: %v", ctx.Err())
+				return ctx.Err()
+			default:
+				resp, err := stream.Recv()
```

Comment:
If there are no messages, which would be normal case most of the time, the ctx.Done() would not be reached, and the context cancelling would not actually work.

Reply:
If there is a chance that a goroutine might go into an infinite loop, it is better to add a check. For the service, this won't be an issue, but for writing a client, it could become a problem🙂

(The function runs in the KeepConnectedToMaster goroutine)

Reply:
"For the service, this won't be an issue" Why is it?

Reply:
There will be no problems for the service, since the `KeepConnectedToMaster` function runs in a single instance. After the application is completed, the `KeepConnectedToMaster` function running in a single instance will also be completed.

If used in the client, the client call can run a lot of `KeepConnectedToMaster` goroutines. If they don't stop after the client completes, then the only option to stop will be to turn off the application (client).

Reply:
Line 243 the `stream.Recv()` is a blocking call. There would be no messages for a long time if there are no messages received, and line 239 `ctx.Done()` would not be reached.

Reply:
I changed it to another check. The context completion check is added at the end of an infinite loop.

https://github.com/seaweedfs/seaweedfs/pull/5628/files#diff-19d64e3838987026ca53e25e1814f47fd552bfa15ed43c1a5743a2916daf3618L172-L259

Reply:
@chrislusf , Done🙂